### PR TITLE
Update plots before scraping

### PR DESCRIPTION
**Pull request by alexrockhill**
Fixes https://github.com/pyvista/pyvista/issues/5277

### CODE REVIEW

**Comment by alexrockhill** [comment]
Can you check if the `plotter` is initialized before calling update? That would fix the errors:

```
FAILED tests/plotting/test_scraper.py::test_scraper[1] - RuntimeError: Render window interactor must be initialized before processing events.
```

Not sure how to do that if anyone has a second.

**Comment by alexrockhill** [comment]
I ran this locally and it fixes my issue.

**Comment by tkoyama010** [comment]
To pass test coverage, we need a sample for adding a test. Please provide us with a sample that you have run locally.

**Comment by tkoyama010** [comment]
pre-commit.ci autofix

**Comment by tkoyama010** [comment]
pre-commit.ci autofix

File: requirements_test.txt [hunk 1]
```diff
@@ -12,11 +12,14 @@ meshio<5.4.0
 nest_asyncio<1.5.9
 numpydoc==1.6.0
 param<2.1.0
+PyQt6!=6.6.1
+PyQt6-Qt6!=6.6.1
```
Comment (tkoyama010) [inline requirements_test.txt:16]:
```suggestion
PyQt6!=6.6.1,<6.7.0
PyQt6-Qt6!=6.6.1,<6.7.0
```

**Comment by alexrockhill** [comment]
Sorry about not figuring out the pre-commit protocol, got it working now and it's very nice.

I'm not sure if it's easiest to mark this to test it only on the CIs with the MNE dependencies. I'm not sure of a way to do that in this case though... `pytest.mark` I don't think would work. Maybe just some logic about trying and failing dependencies. I don't want it to be overly complicated and not look good though.

File: requirements_test.txt [hunk 2]
```diff
@@ -12,11 +12,14 @@ meshio<5.4.0
...
 pytest<7.5.0
 pytest-cov<4.2.0
 pytest-memprof<0.3.0
 pytest-xdist<3.4.0
 pytest_pyvista==0.1.8
+qtpy
```
Comment (tkoyama010) [inline requirements_test.txt:22]:
```suggestion
qtpy<2.5.0
```

**Comment by tkoyama010** [comment]
> Sorry about not figuring out the pre-commit protocol, got it working now and it's very nice.

Never mind. We are happy to show that our CI is excellent.

> I'm not sure if it's easiest to mark this to test it only on the CIs with the MNE dependencies. I'm not sure of a way to do that in this case though... `pytest.mark` I don't think would work. Maybe just some logic about trying and failing dependencies. I don't want it to be overly complicated and not look good though.

I am not as familiar with this bug as you are, but this is a common problem in all environments in my mind. So I'm fine as is.

File: tests/test_tinypages.py [hunk 3]
```diff
@@ -83,3 +84,9 @@ def plot_file(plt, num, subnum):
 
     assert b'you should not be reading this right now' not in html_contents
     assert b'should be printed: include-source with no args' in html_contents
+
+    # test that the plot has the camera position updated with a checksum
+    assert (
+        imread(html_dir / 'plot_cone_00_00.png').sum()
+        != imread(html_dir / 'plot_cone_00_01.png').sum()
```
Comment (tkoyama010) [inline tests/test_tinypages.py:91]:
```suggestion
        imread(html_dir / 'plot_cone_00_01.png').sum()
        != imread(html_dir / 'plot_cone_00_02.png').sum()
```

File: tests/tinypages/plot_cone.py [hunk 4]
```diff
@@ -1,3 +1,12 @@
...
+import pyvista as pv
+
+pl = pv.Plotter()
+pl.iren.initialize()
+pl.app = QApplication.instance()
+pl.add_mesh(pv.Cone())
+pl.camera_position = 'xy'
+
+pl2 = pv.Plotter()
+pl2.add_mesh(pv.Cone())
```
Comment (tkoyama010) [inline tests/tinypages/plot_cone.py:12]:
```suggestion
pv.Cone().plot()

pl = pv.Plotter()
pl.iren.initialize()
pl.app = QApplication.instance()
pl.add_mesh(pv.Cone())
pl.camera_position = 'xy'

pl2 = pv.Plotter()
pl2.add_mesh(pv.Cone())
```

File: tests/test_tinypages.py [hunk 5]
```diff
@@ -83,3 +84,9 @@ def plot_file(plt, num, subnum):
 
     assert b'you should not be reading this right now' not in html_contents
     assert b'should be printed: include-source with no args' in html_contents
+
+    # test that the plot has the camera position updated with a checksum
```
Comment (tkoyama010) [inline tests/test_tinypages.py:88]:
```suggestion
    # test that the plot has the camera position updated with a checksum when the Plotter has an app instance
```

**Review by tkoyama010 — CHANGES_REQUESTED** [review]
Would you mind the output of the image to three pieces in order to keep the existing test intact?

File: .github/workflows/testing-and-deployment.yml [hunk 6]
```diff
@@ -166,6 +166,10 @@ jobs:
           restore-keys: |
             Examples-1-
 
+      - name: Setup xvfb dependencies
+        run: git clone --depth=1 https://github.com/mne-tools/mne-python.git --branch main --single-branch
+        run: ./mne-python/tools/setup_xvfb.sh
```
Comment (tkoyama010) [inline .github/workflows/testing-and-deployment.yml:171]:
```suggestion
        run: |
          git clone --depth=1 https://github.com/mne-tools/mne-python.git --branch main --single-branch
          ./mne-python/tools/setup_xvfb.sh
```

**Comment by alexrockhill** [comment]
Thanks for the review, yes the changes look good

**Review by tkoyama010 — APPROVED** [review]
LGTM. Thanks!

File: tests/plotting/test_scraper.py [hunk 7]
```diff
@@ -16,6 +18,15 @@ def test_scraper(tmpdir, monkeypatch, n_win):
...
     plotters = [pv.Plotter(off_screen=True) for _ in range(n_win)]
+    if n_win == 2:
+        # add cone, change view to test that it takes effect
+        plotters[0].iren.initialize()
+        plotters[0].app = QApplication.instance()
+        plotters[0].add_mesh(pv.Cone())
+        plotters[0].camera_position = 'xy'
+
+        plotters[1].add_mesh(pv.Cone())
+
```
[the remark this answers is not in the document]
Comment (alexrockhill) [reply tests/plotting/test_scraper.py:29]: Sure

**Comment by alexrockhill** [comment]
I got down a bit of a rabbit hole trying to add coverage for the dynamic scraper but it looks like it doesn't have any coverage in general so maybe not the PR to do it in. I was close to adding it, just with `plotters[n].show(interactive=False)` there were 7 vtk objects that we not garbage collected. Might be something to look into as I guess that may cause memory leaks if you use the dynamic scraper?

Should be good to go after implementing the change you asked for with a new test function without branching logic.

File: tests/plotting/test_scraper.py [hunk 8]
```diff
@@ -10,12 +12,61 @@
...
+        plotter.close()
+
+
 @pytest.mark.parametrize('n_win', [1, 2])
 def test_scraper(tmpdir, monkeypatch, n_win):
     pytest.importorskip('sphinx_gallery')
     monkeypatch.setattr(pv, 'BUILDING_GALLERY', True)
     pv.close_all()
     plotters = [pv.Plotter(off_screen=True) for _ in range(n_win)]
+
```
Comment (tkoyama010) [inline tests/plotting/test_scraper.py:69]:
```suggestion
```

File: tests/plotting/test_scraper.py [hunk 9]
```diff
@@ -54,6 +105,7 @@ def test_scraper(tmpdir, monkeypatch, n_win):
     scraper(block, block_vars, gallery_conf)
     for img_fname in img_fnames:
         assert os.path.isfile(img_fname)
+
```
Comment (tkoyama010) [inline tests/plotting/test_scraper.py:108]:
```suggestion
```

**Review by tkoyama010 — APPROVED** [review]
LGTM. Let's make a test of `DynamicScraper `in the future.

File: .github/workflows/testing-and-deployment.yml [hunk 10]
```diff
@@ -166,6 +166,11 @@ jobs:
           restore-keys: |
             Examples-1-
 
+      - name: Setup xvfb dependencies
+        run: |
+          git clone --depth=1 https://github.com/mne-tools/mne-python.git --branch main --single-branch
+          ./mne-python/tools/setup_xvfb.sh
+
```
Comment (banesullivan) [inline .github/workflows/testing-and-deployment.yml:173]: Woah... why was this added here? This needs to be removed ASAP as this section of the tests is strictly non-graphics.

File: requirements_test.txt [hunk 11]
```diff
@@ -12,11 +12,14 @@ meshio<5.4.0
...
 numpydoc==1.6.0
 param<2.1.0
+PyQt6!=6.6.1,<6.7.0
+PyQt6-Qt6!=6.6.1,<6.7.0
 pytest<7.5.0
 pytest-cov<4.2.0
 pytest-memprof<0.3.0
 pytest-xdist<3.4.0
 pytest_pyvista==0.1.8
+qtpy<2.5.0
```
Comment (banesullivan) [inline requirements_test.txt:22]: We do not test Qt dependencies in PyVista. This was a part of the entire motivation for making PyVistaQt a separate repository

**Comment by banesullivan** [comment]
Apologies @alexrockhill and @tkoyama010  but these changes need to be reverted as they are too impactful and need to undergo a more thorough review

File: pyvista/plotting/utilities/sphinx_gallery.py [hunk 12]
```diff
@@ -110,6 +119,10 @@ class DynamicScraper:
 
     """
 
+    def __repr__(self):
+        """Return a stable representation of the class instance."""
+        return f"<{type(self).__name__} object>"
+
```
Comment (banesullivan) [inline pyvista/plotting/utilities/sphinx_gallery.py:125]: @alexrockhill, why was this needed? We do not intend for this class to be used or displayed to users as far as I understand? 

Asking because I'm going to see if I can re-land this PR with less impact

File: .github/workflows/testing-and-deployment.yml (hunk 10 above)
Comment (banesullivan) [reply .github/workflows/testing-and-deployment.yml:173]: Further I want to discourage the use of custom scripts from external projects in our core CI routines. We should keep any configuration and set up limited to well-established GitHub Actions or scripts managed by PyVista.

Adding and executing arbitrary shell scripts is a major red flag and security risk albeit this is from a well-trusted downstream project.

File: pyvista/plotting/utilities/sphinx_gallery.py (hunk 12 above)
Comment (alexrockhill) [reply pyvista/plotting/utilities/sphinx_gallery.py:125]: It was added when I was trying to add coverage of the dynamic scraper but then not removed when adding that coverage exposed a bug where the dynamic scraper failed to close plots and so that was punted to a future PR.

File: .github/workflows/testing-and-deployment.yml (hunk 10 above)
Comment (alexrockhill) [reply .github/workflows/testing-and-deployment.yml:173]: That sounds reasonable. They were added for PyQtx.QWidgets.QApplication.process_events testing. Not sure there is a way to fake that for non-graphics tests, I just don't know there might be though. Otherwise since it was tested it could be no covered.
